feat(web): expand/collapse sidebar

ben-basten opened the pull request:
## Description

<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Adding a collapsible sidebar with a hamburger menu for all breakpoints smaller than Tailwind `md`.

**Benefits include:**

- more space for page content on smaller screen sizes
- seeing the full text of the sidebar buttons on small screens, rather than just seeing the mini sidebar with icons only
- sidebar contains the same content in mobile vs desktop, such as the supporter badge, server info, and online status

**Also in this change:**

- Migrated all buttons in the sidebar and account info panel to use `@immich/ui` components, for consistent spacing and look.
- Moved the theme toggle into the account info panel, to declutter the top navigation. 
- Fix the focus out behavior on the search bar so it will properly activate/deactivate. Currently it stays highlighted and never deactivates after focus leaves.
- Maintain focus in the sidebar if a tree item is clicked on the folders or tags pages
- Use bubbling instead of [capturing](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#usecapture) for the click-outside Svelte action, to allow for finer control of the events, and stopping upward propogation if needed

## How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

- Testing that the sidebar is behaving as expected for all screen sizes in Firefox, Safari, and Chrome on MacOS.
- Focusing the search bar highlights it, removing focus removes the color highlighting
- Theme toggle working in the account info panel
- The sidebar works as expected in the "Administration" settings, Folders page, Tags page, and main Immich sidebar

<details><summary><h2>Screenshots</h2></summary>

<!-- Images go below this line. -->

**New account info panel**

<img src="https://github.com/user-attachments/assets/115432be-9708-453b-ab61-a92d5da1e580" width="200" />

**Sidebar expanded, phone size**

<img src="https://github.com/user-attachments/assets/a77834df-4b92-4a91-8b90-2173b7ddbfec" width="200" />

**Persistent sidebar, tablet size**

<img src="https://github.com/user-attachments/assets/7ffd9a30-e1d4-4716-8e4d-e9ef8500b4a2" width="250" />

**Purchase options, phone size**

<img src="https://github.com/user-attachments/assets/dab52493-394e-462d-9c3a-1189829de249" width="200" />

</details>

<!-- API endpoint changes (if relevant)
## API Changes
The `/api/something` endpoint is now `/api/something-else`
-->

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [x] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code

ben-basten committed:
feat: expand/collapse sidebar

ben-basten committed:
fix: general PR cleanup

- add skip link unit test
- remove unused tailwind styles
- adjust asset grid spacing
- fix event propogation

jrasm91 commented:
Looks like the menu items only work once and then they seem to only trigger closing the sidebar (on mobile at least)

ben-basten committed:
fix: cleaning up event listeners

ben-basten commented:
> Looks like the menu items only work once and then they seem to only trigger closing the sidebar (on mobile at least)

Nice catch @jrasm91, looks like one of the event handlers wasn't being removed properly. It's now fixed in the preview deployment!

ben-basten committed:
fix: purchase modal and button on small screens

ben-basten commented:
I'm curious what you think of the new purchase experience for smaller screen sizes! The purchase options were previously hidden for small screens, so this is all new functionality. I changed the following:

- disabled the hover message when the sidebar is in expand/collapse mode, because it feels cluttered, is not accessible for touch devices, and doesn't position nicely with the overlaid sidebar
- fixed the layout of the purchase modal to be more responsive for smaller screen sizes, collapsing the license options into a column arrangement

jrasm91 commented on web/src/lib/components/elements/buttons/skip-link.svelte:
```diff
@@ -7,24 +7,39 @@
+    if (!breakpoint) {
+      return '';
+    }
+    return `hidden ${breakpoint}:block`;
```
You can't dynamically compute tailwind classes like this. Tailwind scans the source file for class names and patterns and uses the list it finds to create the stylesheet dynamically. You need to have the complete text `sm:block`, etc in the file somewhere.

jrasm91 reviewed (approved):
Very nice!

ben-basten committed:
fix: explicit tailwind classes

ben-basten commented:
The last thing I'm still testing and working on is the "Purchase options" and "About" modals, to get them to play nice with the sidebar.

ben-basten committed:
fix: no animation on initial page load

ben-basten committed:
Merge branch 'main' of https://github.com/immich-app/immich into feat/responsive-sidebar

ben-basten committed:
fix: sidebar spacing and reactivity

ben-basten committed:
Merge branch 'main' into feat/responsive-sidebar

ben-basten committed:
Merge branch 'main' of https://github.com/immich-app/immich into feat/responsive-sidebar

conflict: navigation-bar.svelte

ben-basten commented:
This changes the appearance of the account info panel and moves the dark mode button into the account info panel, which sounds like they might not be desired changes at this time. Let me know if these need to be adjusted, I'm also watching for conversation on #16838.

ben-basten committed:
Merge branch 'main' of https://github.com/immich-app/immich into feat/responsive-sidebar

conflict: purchase-info.svelte

ben-basten committed:
Merge branch 'main' of https://github.com/immich-app/immich into feat/responsive-sidebar

ben-basten committed:
chore: reverting changes to icons in nav and account info panel

ben-basten committed:
Merge branch 'main' of https://github.com/immich-app/immich into feat/responsive-sidebar

conflict: asset-grid.svelte

ben-basten committed:
fix: remove left margin from the asset grid after merging in new timeline

ben-basten committed:
Merge branch 'main' of https://github.com/immich-app/immich into feat/responsive-sidebar

Folding @midzelis top nav changes into the expand/collapse sidebar.

conflicts:
user-page-layout.svelte
navigation-bar.svelte
side-bar-section.svelte

ben-basten committed:
chore: extract search-bar changes for a separate PR

ben-basten committed:
Merge branch 'main' of https://github.com/immich-app/immich into feat/responsive-sidebar

ben-basten committed:
fix: add margin to memories

ben-basten committed:
Merge branch 'main' of https://github.com/immich-app/immich into feat/responsive-sidebar

ben-basten committed:
Merge branch 'main' into feat/responsive-sidebar

alextran1502 commented:
Thank you <3 sorry for the hassle!
